### DISCUSSION TITLE
Running pre-build action while also using internal libraries

### DISCUSSION THREAD

**Discussion by Martinsos** [haskell/cabal#8673] [Q&A] [answered]
Is there any way to run something before building but also use the feature of internal libraries?

We tried using build-type: Custom but then we got error message:
```
Error:
    Internal libraries only supported with per-component builds.
    Per-component builds were disabled because build-type is Custom
```

While when using build-type: Simple, Setup.hs gets ignored.

For context, what we need to run is a shell command that builds a Typescript project, and we want to ensure that always happens before our Haskell program builds.

**Comment by Mikolaj** [comment, marked as answer]
I'd use a makefile, which would also prevent building the Typescript project twice, but perhaps there are better ways.

**Reply by Martinsos** [reply]
And then you would, from that Makefile, both build that Typescript project and also build cabal project, where Typescript project is a dependency of cabal project?

**Reply by Mikolaj** [reply]
I guess so? I'd make it so the parameters to `make` are passed to `cabal` or to `cabal build` inside the Makefile, but I'd keep most of the params in `cabal.project.local` file (`cabal configure` creates such file from the commandline parameters).

BTW, the "prevent building the Typescript project twice" only makes sense if the tool used to build it doesn't take care of that already and if you can tell `make` what file is the output of the corresponding Typescript file so that `make` can verify typestamps.
